### PR TITLE
Slice grammar updates

### PR DESCRIPTION
**Pull request by pepone**
This PR backports the Slice grammar improvements from 865ff422fb4c935726b859d79b73f2788e3add70 and the support for triple-slash doc comments from 7c7a1e8ef665ba8729d5e9e251c251d4bf2c8404

I have to redo this as we have already removed local slice from before_streamline when we did these changes

### CODE REVIEW

**Review by externl — APPROVED** [review]
Looks good to me!

File: cpp/src/Slice/Scanner.l [hunk 1]
```diff
@@ -646,7 +622,7 @@ int checkIdentifier(string& id)
     {
         DefinitionContextPtr dc = unit->currentDefinitionContext();
         assert(dc);
-        if(dc->findMetaData("underscore") != "underscore") // no 'underscore' file metadata
+        if(dc->findMetaData("underscore") != "underscore") // no 'underscore' global metadata
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Scanner.l:625]: I believe 'file' is the correct term still. We moved away from 'global' some time ago.
```suggestion
        if(dc->findMetaData("underscore") != "underscore") // no 'underscore' file metadata
```

File: cpp/src/Slice/Scanner.l [hunk 2]
```diff
@@ -64,546 +41,545 @@ using namespace Slice;
...
+{
+
+void nextLine(int = 1);
+int scanPosition(const char*);
+void setLocation(TokenContext*);
+void startLocation(TokenContext*);
+void endLocation(TokenContext*);
+
+void initScanner();
+void preAction();
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Scanner.l:69]: Looking at the commit that most of this was pulled from (https://github.com/zeroc-ice/ice/commit/865ff422fb4c935726b859d79b73f2788e3add70#diff-3c507567a987cdf989038742a6964ad6d947511ef75ea5ab823136035205722dR67) on line 67.

Was there a reason for removing our custom `fatalError` hookin?
It's definitely not the most important, but in the very rare case there's an internal compiler error, it augments the default error message with some extra information that made debugging much easier for me in the past, before having to ask them to compile with tracing on.

File: cpp/src/Slice/Grammar.y [hunk 3]
```diff
@@ -16,18 +76,17 @@
...
+// warning C4127: conditional expression is constant
+#   pragma warning(disable:4127)
 #endif
 
-//
 // Avoid old style cast warnings in generated grammar
-//
 #ifdef __GNUC__
 #  pragma GCC diagnostic ignored "-Wold-style-cast"
+#  pragma GCC diagnostic ignored "-Wunused-label"
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Grammar.y:86]: I guess we need this warning for the generated Bison code?
It'd be good to update the comment on the enclosing 'ifdef' though.

**Review by InsertCreativityHere — COMMENTED** [review]
I'm still looking through the changes to the grammar file. Everything else was fine.
But am taking a break to finish my enumerator changes first.

File: config/Make.project.rules [hunk 4]
```diff
@@ -178,8 +178,7 @@ endef
 define make-bison
 $1/%.h $1/%.cpp: $1/%.y
 	@$(RM) $$*.h $$*.cpp
-	bison -dvt $2 -o $1/$$*.cpp $$<
```
[the remark this answers is not in the document]
Comment (pepone) [reply config/Make.project.rules:181]: `-t` does the trick https://www.gnu.org/software/bison/manual/html_node/Enabling-Traces.html

`-d` instructs the parser to generate a header file, replaced by `--header` which allows to specify the header file name.

File: cpp/src/Slice/Scanner.l (hunk 2 above)
Comment (pepone) [reply cpp/src/Slice/Scanner.l:69]: Visual Studio reports the function was never used

File: cpp/src/Slice/Scanner.cpp [hunk 5]
```diff
@@ -1580,75 +1829,332 @@ YY_RULE_SETUP
...
+#line 549 "src/Slice/Scanner.l"
+{
+    setLocation(yylloc);
+    return yytext[0];
+}
+	YY_BREAK
+case 49:
+YY_RULE_SETUP
+#line 554 "src/Slice/Scanner.l"
+YY_FATAL_ERROR( "flex scanner jammed" );
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Scanner.cpp:2151]: Maybe VS can't see through the macro, but it really looks like `YY_FATAL_ERROR` is used in the generated code. Here, and about 15 other places.

And in the original changes, we set `#define YY_FATAL_ERROR fatalError`

**Review by InsertCreativityHere — APPROVED** [review]
Looks good!

File: config/Make.project.rules (hunk 4 above)
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply config/Make.project.rules:181]: Ah yes, I got my flags mixed up...

File: cpp/src/Slice/Scanner.cpp (hunk 5 above)
Comment (pepone) [reply cpp/src/Slice/Scanner.cpp:2151]: I will create an issue to fix this one in a separate PR.

**Pull request merged**